### PR TITLE
feat(design): Complete rebuild of the design token package

### PR DESCRIPTION
**Pull request by scotttjob**
## Motivations

- Our design token system has served us well over the years, but it's starting to show it's age.
- The industry has moved towards a standard, to the point where the [w3c is involved](https://tr.designtokens.org/format/)
- We want to meet that standard as closely as we can.
- We want to modernize how we build, and start to offer new funcitonality

## Changes

- Completely rebuilt how we store and generate our design tokens
- Dark Mode is now supported within Storybook
- Tokens are now stored in the [w3c format](https://tr.designtokens.org/format/)
- Now that the tokens are stored in json format, we had to write new code to support them
- We've written a very simple token parser that can generate CSS tokens (--var()), JS tokens {token:val}, css files `:root{}`, js files in both CJS and EJS formats, and theme files.
- The Icons are now using Style tags instead of CSS because of how much easier it is to manage across react and react-native. This is the 'big' change that we could still scale back. We've deleted a LOT of code from the react-native Icon though, so I would love to explore options if we decide this is too big of a change.

### Added

- Dark Mode Toggle Within Storybook
- Completely Rebuilt Design Token System

### Changed

- Design Token System

### Deprecated

- The Old Way of generating design tokens (CSS first)

### Removed

- IconClassNames. This is probably going to be a controversial change, but one I'd like to discuss!

### Fixed

- Lack of a dark mode toggle
- Lack of a dark mode style sheet being served from Atlantis
- Lack of a theming system within Atlantis.

### Security

- 

## Testing

Changes can be
[tested via Pre-release](https://github.com/GetJobber/atlantis/actions/workflows/trigger-qa-build.yml)

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

Random photo of Atlantis

### CODE REVIEW

File: packages/design/rollup.config.mjs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+export default {
+  input: "src/index.ts", // Your entry point
+  output: [
+    {
+      file: "dist/index.mjs", // ES module format
+      format: "es", // ES module format
+    },
+    {
+      file: "dist/index.cjs", // ES module format
+      format: "cjs", // ES module format
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/rollup.config.mjs:15]: ChatGPT, can you please provide me a rollup config file that generates both ejs and cjs, and uses the nodeResolve, commonjs, pluginjson and typescript plugins?

But great catch, they were useless comments and have been removed.

File: packages/design/rollup.config.mjs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import typescript from "@rollup/plugin-typescript";
+import { nodeResolve } from "@rollup/plugin-node-resolve";
+import commonjs from "@rollup/plugin-commonjs"; // Import the plugin
+import pluginjson from "@rollup/plugin-json"; // Import the plugin
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/rollup.config.mjs:4]: ChatGPT in action. They've been removed.

File: packages/design/src/tokens/typography.tokens.json [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      "tight": {
+        "$value": 1.2
+      },
+      "tighter": {
+        "$value": 1.143
+      },
+      "tightest": {
+        "$value": 1.12
+      },
+      "minuscule": {
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/tokens/typography.tokens.json:51]: Great callout and thanks for creating the ticket. I kept seeing the difference and looked up the spelling, just to figure out which is the right one. 

I figure if we're ripping off this design token band-aid, may as well fix a typo at the same time.

File: packages/design/src/tokens/typography.tokens.json [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      },
+      "largest": {
+        "$value": 1.75
+      },
+      "jumbo": {
+        "$value": 2
+      },
+      "extravagant": {
+        "$value": 4
+      },
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/tokens/typography.tokens.json:38]: They exist in mobile I'm fairly certain. So I wondered why we have tokens that exist in one world but not the other.

So I added them to the web world as well. If that turns out to be a mistake, we can walk it back. Felt like one of those small oversights that was easy to fix.

File: packages/design/src/tokens/typography.tokens.json [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      "minuscule": {
+        "$value": 1.08
+      }
+    },
+    "fontSize": {
+      "extravagant": {
+        "$value": "48px"
+      },
+      "jumbo": {
+        "$value": "32px"
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/tokens/typography.tokens.json:60]: Great catch!

File: packages/design/src/maintypography.css [hunk 6]
```diff
@@ -1,53 +0,0 @@
...
-
-  --typography--lineHeight-large: 1.34;
-  --typography--lineHeight-base: 1.25;
-  --typography--lineHeight-tight: 1.2;
-  --typography--lineHeight-tighter: 1.143;
-  --typography--lineHeight-tightest: 1.12;
-  --typography--lineHeight-miniscule: 1.08;
-}
-
-@media (max-width: 639px) {
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/maintypography.css:38]: Agree wholeheartedly that these media queries seem slightly out of place. I definitely think we can migrate them elsewhere!

So the prefix/suffix. It really boils down to the import statements. `imports` needs to come at the top of the file. So we couldn't import our Google Fonts after the design tokens were defined (it throws a warning/error in the browser/build)

But also, some of the CSS we had in this file referenced the design tokens. I THINK the browser is cool if you do that, define tokens after you use them so we could probably just put it all on top?

But it felt weird, so I broke it into two files. One for anything that needs to come before the tokens (imports) and anything that should come after they're defined.

We could rename it to styleImports.css and additionalStyles.css or something less prefix/suffixy as well.

File: packages/design/src/tokens/color.tokens.json [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+      }
+    },
+    "purple": {
+      "$value": "{color.base.pink-.800}"
+    },
+    "purple-": {
+      "dark": {
+        "$value": "{color.base.pink-.900}"
+      }
+    },
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/tokens/color.tokens.json:190]: Yeah I broke the colors up into a few different files, you have the right idea.

baseColor.tokens - Any color token with a direct value. No aliases, this is the base file for everything.
color.tokens - Any token alias that is just a color name. Not semantic, not a base color, not feature specific.
semanticColor.tokens - Semantic colors. Convey meaning instead of color.
workflow.tokens -> Workflow/Feature specific colors, similiar to semanticColors but specific to Jobber Features.

I'm certainly open to other organizations, but that's how I structured it in my head.

File: packages/design/src/tokens/platformOverride.tokens.json [hunk 8]
```diff
@@ -0,0 +1,210 @@
+{
+  "platformOverrides": {
+    "$type": "overrides",
+    "$description": "Platform specific overrides. These are used by the system to override for different platforms. Do not use these directly, but instead add your override value below to have it picked up by the system.",
+    "ios": {
+      "radius": {
+        "circle": {
+          "$value": 100
+        }
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/tokens/platformOverride.tokens.json:9]: Yep! It was easier than trying to juggle it in the transformer.

File: packages/design/package.json [hunk 9]
```diff
@@ -3,56 +3,79 @@
...
     "rollup": "^4.13.0",
-    "rollup-plugin-copy": "^3.5.0",
-    "rollup-plugin-multi-input": "^1.4.1",
-    "rollup-plugin-postcss": "^4.0.2",
-    "rollup-plugin-typescript2": "^0.36.0",
-    "svgo": "^3.0.2",
-    "typed-css-modules": "^0.7.0",
-    "typescript": "^4.9.5"
+    "swc-node": "^1.0.0",
+    "ts-jest": "^29.2.3",
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/package.json:76]: This is strange. So I've been running the tests from the root directory. Just doing `npm run test packages/design` They pass fine.

I tried to get this passing within the directory itself, but ts-jest is having a really hard time using the tsconfig locally. I've left this for now, but I'd love to figure out what's up.

File: packages/design/src/scripts/tokenServices/tokenParsing.ts [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+  return activeType as TokenType;
+};
+
+const handleRecurseError = (e: Error) => {
+  if (e.message.includes("Maximum")) {
+    throw new Error(
+      `Maximum recursion depth reached. Are you missing a $value key in your token structure?`,
+    );
+  } else {
+    console.log("EEEVCCVv", e);
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/scripts/tokenServices/tokenParsing.ts:77]: Nope! Good eye. I've taken it out.

File: packages/design/src/scripts/buildTokens.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    },
+  ]);
+  writeFile("dist/dark.mode.css", darkModeTokens);
+
+  const jsDarkTokens = parseTokens(["dark"], true, "js");
+  const jsTokens = convertRawTokensToJSFile(jsDarkTokens);
+  writeFile("src/assets/tokens.dark.ts", jsTokens);
+
+  const cssTokens = convertRawTokensToCSSFile(cssDarkTokens);
+  writeFile("dist/dark.theme.css", cssTokens);
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/scripts/buildTokens.ts:58]: Yeah I was thinking about how to deliver the tokens in a way that's the most consumable. 

So the dark.theme.css is just the color tokens, applied to the root. dark.mode.css is the exact contents of the file that we would need within JO.

We certainly might not need dark.theme.css initially (or ever) I might be prematurely guessing what people could possibly want. I also think I I did this right near the end as well, so I might have been rushing a bit.

File: packages/design/src/scripts/tokenServices/tokenConstants.ts [hunk 12]
```diff
@@ -0,0 +1,2 @@
+export const baseUnit = "16px";
+export const MAX_DEPTH = 200;
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/scripts/tokenServices/tokenConstants.ts:2]: Done! At some point I was manually keeping track of recursion depth to prevent the error, but then I figured it just made more sense to let the error happen and catch it instead.

File: packages/design/src/scripts/buildIconMap.ts [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+      (path: SVGPathElement) =>
+        (path.attributes as unknown as { d: { value: string } }).d.value,
+    );
+    iconMap[file.replace(".svg", "")] = pathAttributes;
+  }
+
+  return iconMap;
+};
+
+export const getIconMap = async () => {
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/scripts/buildIconMap.ts:25]: Done!

File: packages/design/src/scripts/buildIconMap.ts [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    const iconMap = await processFiles(files, iconDir);
+
+    return { icons: iconMap };
+  } catch (err) {
+    console.error("Error reading the directory or files:", err);
+    throw err; // Rethrow or handle as needed
+  }
+};
+
+export const generateIconMapFile = () => {
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/scripts/buildIconMap.ts:40]: Done!

File: packages/design/src/tokens/space.tokens.json [hunk 15]
```diff
@@ -0,0 +1,33 @@
+{
+  "space": {
+    "$type": "dimension",
+    "$description": "Space tokens. These are used to define the space between elements. Feel free to use these directly, but only as a space token. Do not use these as a dimension token.",
+    "minuscule": {
+      "$value": "1px"
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/tokens/space.tokens.json:6]: We used them with the idea that someone could adjust 'base-unit' in a css custom property and adjust how all the styles within adjust accordingly. So for instance someone could shrink the base-unit on a wrapper, and all the spacing would condense.

The main problem with it now is that calc() is not part of the W3C spec, we had to employ a fairly expensive postcss step (in both the old and new ways) specifically to calculate these values for the react-native/javascript side.

So I figured I would remove the calcs, and if we wanted to bring the functionality back to the web/JO, we could maybe add these calc statements to the suffixStyles.css file, so we didn't have to employ the expensive postcss step just to make them react-native compatible.

Basically move the platform-specific implementation closer to the platform.

I chatted with Chris briefly before his vacay, but I think this one definitely warrants a bit of a conversation.

File: packages/design/src/tokens/workflow.tokens.json [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+      "$value": "{color.base.blue-.800}"
+    },
+    "task-": {
+      "surface": {
+        "$value": "{color.base.blue-.200}"
+      },
+      "onSurface": {
+        "$value": "{color.base.blue-.900}"
+      }
+    },
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/tokens/workflow.tokens.json:48]: Fixed! I'm wondering if these were changed recently (or I was just sloppy on the copying)

File: packages/design/src/iconStyles/iconStyles.ts [hunk 17]
```diff
@@ -0,0 +1,112 @@
+export const iconStyles: Record<string, Record<string, string>> = {
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/iconStyles/iconStyles.ts:1]: Done!

File: packages/design/src/iconStyles/iconColours.ts [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+    },
+    warningOnSurface: {
+      value: "{color.warning-.onSurface}",
+    },
+    success: {
+      value: "{color.success}",
+    },
+    successOnSurface: {
+      value: "{color.success-.onSurface}",
+    },
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/iconStyles/iconColours.ts:89]: Done!

File: packages/design/src/tokens/platformOverride.tokens.json [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+          },
+          "smaller": {
+            "$value": 12
+          },
+          "smallest": {
+            "$value": 10
+          }
+        },
+        "lineHeight": {
+          "large": {
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/tokens/platformOverride.tokens.json:150]: Done!

File: packages/design/src/scripts/tokenServices/tokenTypes.ts [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+  string,
+  string | { $value: string } | object | number
+>;
+export interface OverrideTokenTree {
+  platformOverrides: Record<"ios" | "android", TokenTree>;
+}
+export type Token = string | object | number;
+export type Tokens = Record<string, Token>;
+export type ParsedTokens = Record<string, string | number>;
+export type ParsedTokenValue = string | number;
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/scripts/tokenServices/tokenTypes.ts:20]: Done! Nice catch. Probably a case of me creating one, then the other, and then not going back to clean up

File: packages/design/src/scripts/buildTokens.ts [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+
+  const iosTokens = createTokenFileContentsForPlatform("ios");
+  writeFile("src/assets/tokens.ios.ts", iosTokens);
+};
+
+const writeColorTokens = () => {
+  const jsFileContents = createTokenFileFromSubset(["color"]);
+  writeFile("src/assets/tokens.color.ts", jsFileContents);
+  writeFile("dist/colors.mjs", jsFileContents);
+  writeFile("dist/colors.cjs", jsFileContents);
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/scripts/buildTokens.ts]: hahaha regular token in that it's part of the token design structure. I'm just appending the value into the structure when generating it now.

File: packages/design/package.json [hunk 22]
```diff
@@ -3,56 +3,79 @@
...
-    "build:iconCssStyleTypes": "tsc --declaration --emitDeclarationOnly --allowJs src/icons/iconStyles.ts --outDir ./src/icons --skipLibCheck",
-    "clean": "rm -rf dist/* icons/* src/icons/iconMap.ts colors.js foundation.* iconStyles.* tsconfig.tsbuildinfo",
+    "test": "jest",
+    "build:design": "rollup -c rollup.config.mjs",
+    "generate:iconMap": "tsx src/scripts/buildIconMap.ts",
+    "generate:tokenCSS": "tsx src/scripts/buildTokens.ts",
+    "generate:assets": "npm run generate:iconMap && npm run generate:tokenCSS",
+    "build": "mkdir -p dist && npm run generate:assets && npm run build:design",
+    "watch": "mkdir -p dist & chokidar \"**/*.ts\" \"**/*.json\" \"**/*.svg\"-c \"npm run generate:assets && npm run build:design\"",
+    "clean": "rm -rf dist colors.js foundation.css src/assets colors.cjs colors.mjs node_modules",
```
Comment (MichaelParadis) [inline packages/design/package.json:59]: I am not sure if we want to delete `node_modules` since lerna's clean will do that and the clean script runs in prepare so node_modules are nuked in the initial `npm ci`
